coderedcms/models/page_models.py

```diff
@@ -1394,18 +1431,61 @@ def serve(self, request, *args, **kwargs):
         )
         return response
 
-    preview_modes = [
-        ('form', _('Form')),
-        ('landing', _('Thank you page')),
-    ]
+class CoderedSessionFormSubmission(SessionFormSubmission):
+    class Meta:
+        proxy=True
 
-    def serve_preview(self, request, mode):
-        if mode == 'landing':
-            request.is_preview = True
-            return self.render_landing_page(request)
+    def get_flattened_data(self):
+        flattened_data = {}
+        for step in self.get_data():
+            for k, v in step.items():
+                flattened_data[k.replace('-', '_')] = v
+        return flattened_data
 
-        return super().serve_preview(request, mode)
+class CoderedAdvancedFormPage(StreamFormMixin, CoderedFormMixin, CoderedWebPage):
+    class Meta:
+        verbose_name = _('CodeRed Advanced Form Page')
+        abstract = True
+
+    template = 'coderedcms/pages/advanced_form_page.html'
+    landing_page_template = 'coderedcms/pages/form_page_landing.html'
+
+    form_fields = StreamField([('step', CoderedAdvancedFormStepBlock()),])
+    encoder = StreamFormJSONEncoder
+
+    body_content_panels = [
+        StreamFieldPanel('form_fields')
+    ] + \
+    CoderedFormMixin.body_content_panels + [
+        InlinePanel('confirmation_emails', label=_('Confirmation Emails'))
+    ]
 
+    def serve(self, request, *args, **kwargs):
+        context = self.get_context(request)
+        form = context['form']
+        if request.method == 'POST' and form.is_valid():
+            is_complete = self.steps.update_data()
+            if is_complete:
+                submission = self.get_submission(request)
+                self.process_form_submission(
+                    request=request,
+                    form=form,
+                    form_submission=submission,
+                    processed_data=submission.get_data()
+                )
+                return self.serve_success(request, *args, **kwargs)
+            return HttpResponseRedirect(self.url)
+        return CoderedWebPage.serve(self, request, *args, **kwargs)
+
+    @classmethod
+    def get_submission_class(cls):
+        return CoderedSessionFormSubmission
+
+    def get_storage(self):
```

Comment:
Isn't this already provided via the mixin?

Reply:
Yeah, but it then gets overwritten again by the `StreamFormMixin`...I'll have to think of a cleaner solution.